src/wp-includes/script-loader.php

```diff
@@ -2830,9 +2830,18 @@ function wp_print_script_tag( $attributes ) {
  * @return string String containing inline JavaScript code wrapped around `<script>` tag.
  */
 function wp_get_inline_script_tag( $javascript, $attributes = array() ) {
-	if ( ! isset( $attributes['type'] ) && ! is_admin() && ! current_theme_supports( 'html5', 'script' ) ) {
+	$is_html5 = current_theme_supports( 'html5', 'script' ) || is_admin();
+	if ( ! isset( $attributes['type'] ) && ! $is_html5 ) {
 		$attributes['type'] = 'text/javascript';
 	}
+
+	// Ensure markup is XHTML compatible if not HTML5.
+	if ( ! $is_html5 ) {
+		$javascript = sprintf( "/* <![CDATA[ */\n%s\n/* ]]> */", $javascript );
```

Comment:
To ensure XML-compatibility, the `$javascript` string should have any instances of `]]>` escaped. It's ugly, but apparently this is what has to be done:

```suggestion
		$javascript = str_replace( ']]>', ']]]]><![CDATA[>', $javascript );
		$javascript = sprintf( "/* <![CDATA[ */\n%s\n/* ]]> */", $javascript );
```

Nevertheless, it is likely _exceedingly_ rare that a WordPress site is actually being served with `Content-Type: application/xhtml+xml` (true confessions, I used to do it, but the draconian XML parse error handling was painful). Still, it is unlikely for `]]>` to occur in a script. I say this because if a site is being served as `Content-Type: text/html` then the parser HTML parser will ignore these `CDATA` sections, and it could end up causing a parse error when the script is passed to the JS interpreter.

Reply:
Is there any concern from adding this now in case the `$javascript` passed to this function is already wrapped with that `CDATA` markup? Very unlikely I assume, but we may want to add a check?

Reply:
Yes, good call. I've done this in ecc29a9. This shows that it is indeed necessary to do the escaping. Without the escaping, doing the following:

```php
wp_print_inline_script_tag( "/* <![CDATA[ */ console.log( 'Hello World!' ); /* ]]> */" );
```

Results in the following output:

```xml
<script type="text/javascript">
/* <![CDATA[ */
/* <![CDATA[ */ console.log( 'Hello World!' ); /* ]]> */
/* ]]> */
</script>
```

Pasting this into the W3C Validator as a fragment under the XHTML Strict doctype:

> ![image](https://github.com/WordPress/wordpress-develop/assets/134745/fbbd65e5-bb66-427f-85ea-95317c2eb67a)

Results in an XML parse error:

> ![image](https://github.com/WordPress/wordpress-develop/assets/134745/42a369a0-6fe4-4e2d-986d-4e15f9feb6f2)

However, when the escaping is present, the PHP outputs:

```html
<script type="text/javascript">
/* <![CDATA[ */
/* <![CDATA[ */ console.log( 'Hello World!' ); /* ]]]]><![CDATA[> */
/* ]]> */
</script>
```

And this is valid:

> ![image](https://github.com/WordPress/wordpress-develop/assets/134745/037c528a-2392-4f2b-9415-063379a0e64e)

Reply:
something as bizarre as this could definitely use a link to an XML spec or the source of the reason why these have to be escaped. someone is going to look at that and "fix" it and "improve quality" by removing the escaping 🙃

Reply:
Even though there is a comment about the escaping?

Reply:
yeah because it mentions _that_ it should be compatible but not _how_ or _why_.
like, _how does this fix ensure compatibility?_

when I saw it I immediately wondered what rule necessitates this or what breaks without it. I could imagine something like this, to see if I'm properly understanding the code.

```php
/*
 * XHTML extracts the contents of the SCRIPT element and then the XML parser
 * decodes character references and other syntax elements. This can lead to
 * misinterpretation of the script contents or invalid XHTML documents.
 *
 * Wrapping the contents in a CDATA section instructs the XML parser not to
 * transform the contents of the SCRIPT element before passing them to the
 * JavaScript engine.
 *
 * Example
 *
 *     <script>console.log('&hellip;');</script>
 * 
 *     In an HTML document this would print "&hellip;" to the console,
 *     but in an XHTML document it would print "…" to the console.
 * 
 *
 *     <script>console.log('An image is <img> in HTML');</script>
 * 
 *     In an HTML document this would print "An image is <img> in HTML",
 *     but it's an invalid XHTML document because it interprets the `<img>`
 *     as an empty tag missing its closing `/`.
 *
 * @see https://www.w3.org/TR/xhtml1/#h-4.8
 */
if ( ! $is_html5 ) {
	/*
	 * If the string `]]>` exists within the JavaScript it would break
	 * out of any wrapping CDATA section added here, so to start it's
	 * necessary to escape that sequence, which requires splitting the
	 * content into two CDATA sections wherever it's found.
	 *
	 * Note: it's only necessary to escape the closing `]]>` because
	 * an additional `<![CDATA[` leaves the contents unchanged.
	 */
	$javascript = str_replace( ']]>', ']]]]><![CDATA[>', $javascript );

	// Wrap the entire escaped script inside a CDATA section.
	$javascript = sprintf( "/* <![CDATA[ */\n%s\n/* ]]> */", $javascript );
}
```

yeah I know it's wordy and lengthy, but it was really confusing to me, and this code is now going to be responsible for general code generation, and may get a lot of eyes on it.

by the way it seems like this _will not_ trigger if the HTTP `Content-type: text/html` is what serves the document. I could not uncover these failures without serving as `Content-type: application/xhtml+xml` or by directly opening the file with a `.xml` extension. The exact same file contents stored with `.html` as its extension leads to HTML semantics, thus this doesn't effect it.

Finally, and thankfully, it doesn't appear to be a security issue because Safari, Firefox, and Chrome all prevent escaping from the script using tricks like `&lt;/script&gt;`. It will change those things into `</script>`, but that leads to a JavaScript syntax error, or data corruption if it's found within a JS string.

Including `<img>` was fun because that broke the entire page render. It became invalid XML. Again, if served as `.html` or with `Content-type: text/html` or anything but explicit out-of-band information that it's XML, none of these problem arise.